src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -488,49 +501,41 @@ def select(self, cols: List[Expression]) -> "SelectStatement":
             # We don't flatten when there are duplicate columns.
             can_be_flattened = False
             disable_next_level_flatten = True
-        elif self.flatten_disabled or self.has_clause_using_columns:
+        elif self.flatten_disabled:
+            can_be_flattened = False
+        elif (
+            self.where
```

Comment:
Core change #2: Previously, as long as subquery has a WHERE clause, `can_be_flattened` evaluates to `False`. This change deems that the only case not to be flattened is when any dependent column from the WHERE clause is newly added in the current query, i.e. `df.filter(col("C") > 10).select("*", col(Literal(12)).alias("C")` should not be flattened to `SELECT *, 12 as C FROM <subquery> WHERE C > 10`.

Reply:
This is due to my lack of knowledge of the simplifier: Do we handle the case where select added alias to existing column? E.g. `df.filter(col("C") > 10).select(col("A").alias("C"))`? Is this alias also considered a new column? If so then I think it should be okay.

Reply:
We don't flatten `df.filter(col("C") > 10).select(col("A").alias("C"))`. `C` is considered a new column in this case.